pkg/resource/deploy/providers/registry.go

```diff
@@ -0,0 +1,322 @@
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check the provider's config. If the check fails, unload the provider.
+	inputs, failures, err := provider.CheckConfig(olds, news)
+	if len(failures) != 0 || err != nil {
+		closeErr := r.host.CloseProvider(provider)
+		contract.IgnoreError(closeErr)
+		return nil, failures, err
+	}
+
+	// If we are running a preview, configure the provider now. If we are not running a preview, we will configure the
+	// provider when it is created or updated.
+	if r.isPreview {
+		if err := provider.Configure(inputs); err != nil {
+			closeErr := r.host.CloseProvider(provider)
+			contract.IgnoreError(closeErr)
+			return nil, nil, err
+		}
+	}
+
+	// Create a provider reference using the URN and the unknown ID and register the provider.
+	r.setProvider(mustNewReference(urn, UnknownID), provider)
+
+	return inputs, nil, nil
+}
+
+func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.PropertyMap,
+	allowUnknowns bool) (plugin.DiffResult, error) {
+
+	contract.Require(id != "", "id")
+
+	label := fmt.Sprintf("%s.Diff(%s,%s)", r.label(), urn, id)
+	logging.V(7).Infof("%s: executing (#olds=%d,#news=%d)", label, len(olds), len(news))
+
+	// Create a reference using the URN and the unknown ID and fetch the provider.
+	provider, ok := r.GetProvider(mustNewReference(urn, UnknownID))
+	contract.Assertf(ok, "'Check' must be called before 'Diff'")
+
+	// Diff the properties.
+	diff, err := provider.DiffConfig(olds, news)
+	if err != nil {
+		return plugin.DiffResult{Changes: plugin.DiffUnknown}, err
+	}
+
+	// If the diff requires replacement, unload the provider: the engine will reload it during its replacememnt Check.
+	// If the diff does not require replacement and we are running a preview, register it under its current ID.
+	if len(diff.ReplaceKeys) != 0 {
+		closeErr := r.host.CloseProvider(provider)
+		contract.IgnoreError(closeErr)
+	} else if r.isPreview {
+		r.setProvider(mustNewReference(urn, id), provider)
+	}
+
+	return diff, nil
+}
+
+func (r *Registry) Create(urn resource.URN,
+	news resource.PropertyMap) (resource.ID, resource.PropertyMap, resource.Status, error) {
+
+	contract.Assert(!r.isPreview)
+
+	label := fmt.Sprintf("%s.Create(%s)", r.label(), urn)
+	logging.V(7).Infof("%s executing (#news=%v)", label, len(news))
+
+	// Fetch the unconfigured provider, configure it, and register it under a new ID.
+	provider, ok := r.GetProvider(mustNewReference(urn, UnknownID))
+	contract.Assertf(ok, "'Check' must be called before 'Create'")
+
+	if err := provider.Configure(news); err != nil {
+		return "", nil, resource.StatusOK, err
+	}
+
+	id := resource.ID(uuid.NewV4().String())
+	contract.Assert(id != UnknownID)
+
+	r.setProvider(mustNewReference(urn, id), provider)
+	return id, resource.PropertyMap{}, resource.StatusOK, nil
+}
+
+func (r *Registry) Read(urn resource.URN, id resource.ID,
+	props resource.PropertyMap) (resource.PropertyMap, error) {
+	contract.Fail()
```

Comment:
what's stopping me from creating a custom resource whose type is `pulumi:providers:MyProvider` and trying to read it?